### PR TITLE
feat(og-gen): Implement middleware and hooks 

### PR DESCRIPTION
**Pull request by dac09**
The OG Gen saga continues with  @cannikin and @dac09 ⚔️

This PR adds OgImageMiddleware and Hooks to `@redwoodjs/og-gen`.

This PR:
- adds OgImageMiddleware and Hooks to `@redwoodjs/og-gen`, complete with tests

⚠️ Template changes:
- updates entry.client template to pass in Routes to App
- updates App to take children (i.e. Routes)

This is so that we can pass the OG component to be rendered _with_ your App's CSS setup.
I've also coordinated with Tobbe on these changes so we don't need to change things again for server router. 

**How to use this?**

1. **Registering the middleware:**
    ```ts 
    import OgImageMiddleware from '@redwoodjs/ogimage-gen/middleware'

    export const registerMiddleware = () => {
      const ogMw = new OgImageMiddleware({
        App,
        Document,
      })

      return [ogMw]
    }
    ```

2. Configure your `vite.config.ts`
    ```ts
    import vitePluginOgImageGen from '@redwoodjs/ogimage-gen/plugin'

    const viteConfig: UserConfig = {
      // 👇 so it builds your OG components
      plugins: [redwood(), vitePluginOgImageGen()],
    }

    export default defineConfig(viteConfig)
    ```
3. Add your OG Image component next to the page it's for
e.g. web/src/pages/AboutPage/AboutPage.png.tsx

4. Use hooks on AboutPage to generate the ogURL

**Tasks:**
- [x] Can we write unit tests for OgImageMiddleware?
- [x] I haven't tried the hooks yet! RC maybe try out on your project?
- [x] Move hooks from Rob's project to this PR
- [x] Figure out ESM/CJS issue we're seeing
- [x] Find those pesky edge cases e.g. multiple .png.jsx files in a page folder
~- [ ] Can we support image src imports? Not strictly necessary.~
Known limitation. You have to put images in the public folder. We can improve later
~- [ ] DOCS~
We decided to make it a blog post instead, since this is under experimental still

### CODE REVIEW

**Review by Tobbe — CHANGES_REQUESTED** [review]
There are some TODOs in here that I'd be fine with leaving around. But some that I definitely want you to take a second look at before merging. Do you want me to comment on all of them now, or should I hold off for a bit until you're further along with PR and ready for a proper review?

**Comment by dac09** [comment]
> There are some TODOs in here that I'd be fine with leaving around. But some that I definitely want you to take a second look at before merging. Do you want me to comment on all of them now, or should I hold off for a bit until you're further along with PR and ready for a proper review?

Whatever you like! 

The TODO is mostly for Rob though.

**Comment by Tobbe** [comment]
I’ll wait until Rob has had a chance to address them then 🙂

File: packages/ogimage-gen/src/OgImageMiddleware.ts
Comment (dac09) [inline packages/ogimage-gen/src/OgImageMiddleware.ts:213]: @Tobbe you could help with this one please haha

File: packages/ogimage-gen/src/OgImageMiddleware.ts
Comment (cannikin) [inline packages/ogimage-gen/src/OgImageMiddleware.ts:159]: So in entry.server you'd pass App, Document and playwright to new OgImageMiddleware()? Hmm...does it effect bundle size? Or does it just feel better that you're passing the dependency instead of having this class require import it itself?

If the rendering code was somehow generic enough that it could use any renderer, and playwright is just one possibility, I think that would be awesome: you pass renderer at instantiation time and it just uses that to render the image, but there isn't currently any other option, is there? Didn't our testing show that only Playwright was able to render everything correctly (since it was a real web browser and parsing CSS the same way, unlike that Vercel package which supported a subset of CSS)?

File: packages/ogimage-gen/src/OgImageMiddleware.ts
Comment (cannikin) [inline packages/ogimage-gen/src/OgImageMiddleware.ts:145]: Can confirm! It looks like this is because the `relativeFilePath` is the path to the page itself, and that's still a .tsx file (I only renamed the og:image component to `.jsx`). So it looks for the same named file but with `.png.tsx` on the end, which doesn't exist. If you rename the page itself to also be `.jsx` then it works! So they both need to be `.jsx` or both `.tsx`.

Personally I'm okay with this limitation, I feel like most people are going to consistently have one or the other. If we wanted to be nice we could try/catch that error and say "Your og:image component needs to have the same extension as its parent page!", otherwise we'd have to do some file system shenanigans to try and be smart and convert one to the other, ugh.

Comment (dac09) [reply packages/ogimage-gen/src/OgImageMiddleware.ts:145]: Will remove the TODO

File: packages/ogimage-gen/src/OgImageMiddleware.ts
Comment (dac09) [reply packages/ogimage-gen/src/OgImageMiddleware.ts:159]: > So in entry.server you'd pass App, Document and playwright to new OgImageMiddleware()?

Yeah exactly this. It does not affect bundle size, because this is all on the server. It's mainly a way of making it clear that the user will need to install playwright, then pass it to us. So both `yarn install playwright` and `npx playwright install` isn't really a middleware setup step then, more something they need to do on their project. Maybe I'm just overcomplicating things though. 

> If the rendering code was somehow generic enough
haha lets not go there.... but yes a possibility in the future.

Comment (dac09) [reply packages/ogimage-gen/src/OgImageMiddleware.ts:159]: See readme for playwright on how to install. https://www.npmjs.com/package/playwright

I suppose we have the same in OgMiddleware? To use you need to run this stuff?

Comment (cannikin) [reply packages/ogimage-gen/src/OgImageMiddleware.ts:159]: Yeah that makes sense. After running `yarn rw setup middleware ogimage` we can have a simplified version of those install instructions in the last step.

**Comment by dac09** [comment]
@Tobbe ready for once over.

And for final checks @cannikin ✌️

**Comment by dac09** [comment]
FINALLY. All checks ✅

Will wait for docs @cannikin, but not sure if we're putting it in this PR. Will leave to you for merge. 

@Tobbe any nits by chance? Could you please update your review?

**Comment by Tobbe** [comment]
I’ll have a look

File: .changesets/10469.md
Comment (Tobbe) [inline .changesets/10469.md:1]:
```suggestion
- feat(og-gen): Implement middleware and hooks  (#10469) by @dac09
```

File: .changesets/10469.md
Comment (Tobbe) [inline .changesets/10469.md:13]:
```suggestion
```

**Review by Tobbe — APPROVED** [review]

**Comment by cannikin** [comment]
I built an app from scratch using this branch and everything worked great! I manually made the changes to `entry.server.js` and `vite.config.js` but @Josh-Walker-GM has a PR that adds CLI commands to modify them automatically: #10485

**Comment by cannikin** [comment]
No official docs for the real site yet, but we'll have an announcement blog post that walks through the whole thing. I think when we get closer to SSR being promoted from experimental to feature flag I can write an official reference doc for it.

**Pull request merged**
